### ISSUE TITLE
NNI 2021 Oct~Nov Iteration Plan

### ISSUE THREAD

**Issue by scarlett2018** [microsoft/nni#4211]
This is the plan for the iteration in 2021 Nov~Dec Iteration, it's a 6 weeks iteration.

# Release Plan for upcoming release
- Release manager: @liuzhe-lz
- Feature freeze date: ~~11-19~~ 11-26 (last delay)
- Code freeze and demo date: ~~12-3~~ ~~12-8~~ 12-10 (last delay)
- Branch cut and next release planning date: ~~12-7~~ 12-24
- Release date: ~~12-10~~ 12-29

### Documentation Update

- [x] @ultmaster Check statics
  - [x] @QuanluZhang Get Ni Yan's account
- [x] Decide which ones to keep Chinese doc
- [x] Update docs for current release
- [x] Deal with Chinese doc @ultmaster (12.1) #4345
- [x] HPO doc hierarchy @ zebin (11.23)

### NAS

- end2end example for NAS with Retiarii (**target starting date 11/17**)
   - [x]  P0 - @JiahangXu evaluate current SPOS support https://nni.readthedocs.io/en/stable/NAS/SPOS.html #4322
   - [x]  P0 - @JiahangXu evaluate current ProxylessNAS support https://nni.readthedocs.io/en/stable/NAS/Proxylessnas.html (11.19) (Jiahang: the example works, refine its doc in PR #4321 )
- [x] P0 - @ultmaster Serialization (~~11.5~~ ~~11.12~~ ~~11.19~~ 11.24) #4337
- P1 - @QuanluZhang Fix webui chart
  - [x] survey (11.3)
  - [x] fix (11.10) #4304
  - [x] further design
- [x] P1 - @Lijiaoa Trial parameters (11.23) #4314
- P1 - @QuanluZhang ~~option of not exit experiment after done~~ #4340

### Model Compression

- [x] P0 - @J-shang Add doc for v2 config list
- Pruner Improvement
  - [x] @J-shang Support pruning by step (movement pruning, taylor, activation...). (11.12) #4308
- Migrate Pruner from v1
  - [x] @J-shang AutoCompress Pruner. (11.5) #4280 review: @QuanluZhang @Fiascolsy 
  - [x] P0 - @Fiascolsy AMC Pruner. (~~11.5~~ ~~11.12~~ 11.19) #4320
- [ ] P2 - @Fiascolsy Fix AutoCompressPruner GPU memory leak.
- Doc
  - [ ] @J-shang Guide on how to customize pruner in v2.
  - [ ] P1 - @Fiascolsy Report performance of v2 compare with paper / v1. #4365
  - [ ] Need add: fine-grained with speed up in iteration will case issue. how to write config_list.
  - review doc & update the outdated doc.
    - [ ] P1 - @Fiascolsy pruning
    - [ ] @linbinskn quantize
  - [x] P1 - @J-shang review overview and tutorial hirachey
- [x] P2 - @Fiascolsy More v2 Example #4305
- [x] @J-shang pass optimizer to pruner redesign (11.26) #4332

### HPO
- [ ] @QuanluZhang Survey optuna, onedb, sagemaker
- [ ] @QuanluZhang fix dispatcher error, experiment status
- [x] P1 - (zebin) improve tuner doc
- [ ] @ultmaster Survey ConfigSpace
- [x] P1 - @liuzhe-lz TPE (PR #4239)
   - [x] doc #4339
- Nested search space
  - [x] P1 - @liuzhe-lz Grid search (~~11.5~~ ~~11.10~~ 11.19) #4324

### nnictl
- [x] P0 - @liuzhe-lz Deprecate config v1 (~~11.1~~)
  - [x] `ConfigBase` class #4303
  - [x] local training service #4313
  - [x] All training services #4318
  - [x] launcher #4330
  - [x] update nni manager #4355
  - [x] fix annotation (11.26)
  - [x] run ITs
- [x] P1 - @liuzhe-lz Raise error for setting invalid field on ConfigBase
- [x] P0 - @liuzhe-lz add validation for trialGpuNumber/maxTrialNumberPerGpu combination

### NNI manager & training service

- [x] P1 - @SparkSnail Pipeline for reuse training services (~~11.5~~ ~~11.8~~ 11.10) #4310
- [x] P1 - @SparkSnail Hybrid for all reuse training services (11.26) #4343

### WebUI

- [x] Overview page: blank at the bottom (11.16)
- [x] Update readthedoc index.html (~~11.4~~ 11.10)
- ~~About trial duration question~~ https://github.com/microsoft/nni/issues/3012

### Other

- readthedocs dependency
- doc translation

### NAS

- [ ] P1 - @JiahangXu a TBD end2end example for NAS with Retiarii
- [ ] P1 - @QuanluZhang ~~Support tuner_no_more_trial~~
- [x] P1 - @JiahangXu review graph engine
- [ ] P? - @ultmaster Update mnasnet and example refactor (plan: #4249)
- [ ] P2 - (zebin) More choice type design
- [ ] P2 - (zebin) Survey naslib
- [ ] P2 - (zebin) Port Houwen's algorithm
- [ ] P2 - @liuzhe-lz Merge `RetiariiExperiment` back to `Experiment`
- [ ] P3 - Visualization nasbench
- [ ] P3 - Support HPO in Retiarii
- [ ] P3 - Review one-shot, refactor with lightning

### Compression

- Pruning Speed Up (check with ningxin)
- [ ] P1 - Fine-grained iterative speedup logic.
- [ ] P1 - NetAdapt Pruner
- [ ] @J-shang Transformer Pruner. (~~11.5~~ 11.8) #4180
- [ ] @zheng-ningxin Function replacement
- [ ] P1 - Sensitivity Pruner.
- [ ] @J-shang ADMM: support more metrics.
- [ ] P2 - Support min_sparsity_per_layer in global mode.
- [ ] P2 - Mixed sparse pattern.
- [ ] P2 - Support mask output/input.
- [ ] P2 - Pruning with constraints (remain 4x channels).
- [ ] P2 - @Fiascolsy Iterative pruning starts from any sparsity ratio.
- [ ] P2 - KD in pruning scheduler (maybe use original model as teacher).
- [ ] @zheng-ningxin Memory usage optimization
- [ ] Refactor of model graph generation (as an independent component)
- [ ] Speed up with special/customized layer.
- Quantization V2

### HPO

- [ ] P1 - fix `len(arr) > 0` problem
- [ ] P1 - (zebin) Improve hyperband & PBT with shared storage
- [ ] P3 - Multi-objective
- [ ] P2 - (zebin) transferable HPO
- [ ] P2 - Evolution nested search space

### nnictl

- [ ] P2 - @J-shang Improve `nnictl experiment export`
- [ ] P2 - @liuzhe-lz auto-completion
- [ ] P2 - @liuzhe-lz early validation

### NNI Manager

- [ ] P2 - Distributed trial
- [ ] P1 - @acured Project back-end

### Web UI

- [ ] Align issue in overview page
- [ ] Support searching customized string in the search box
- [ ] Experiment list could not open experiment with prefix https://github.com/microsoft/nni/issues/3746 (need backend)

**Comment by chenbohua3** [comment]
Hi @scarlett2018 I want to add these items for this iteration:
+ About quantization:
  + Support dtype & qscheme customization for lsq quantizer and observer quantizer (P0)
  + upgrade lsq quantizer to lsq+ quantizer, which can used for asymmetric quantization (P0)
  + add model converter for mnn backend (P1)

+ About architecture for quantization V2:
  + Analyze the architecture of the PyTorch quantization tool in the open source community, and summarize the advantages and disadvantages of each tool.(P0), with @linbinskn 

+ A demo case for  quantizing YOLOX and deploy it on PyTorch backend (P1)

**Comment by liuzhe-lz** [comment]
#4149 @zheng-ningxin (discuss meeting)
#4239 @QuanluZhang (update)
#4314 @QuanluZhang
#4308 @QuanluZhang (update) @zheng-ningxin
#4307 @QuanluZhang (update)  @Fiascolsy (update)
#4305 @QuanluZhang (update)  @J-shang (update)
#4314 @QuanluZhang @liuzhe-lz
#4318 @J-shang (update) @sparksnail (update)
#4319 @scarlett2018
#4321 @ultmaster @QuanluZhang @liuzhe-lz
#4325 @ultmaster @QuanluZhang @liuzhe-lz
#4328 @J-shang @liuzhe-lz

**Comment by liuzhe-lz** [comment]
PRs to be merged before Code Freeze:

- [ ] #4304
- [ ] #4314
- [ ] #4320
- [ ] #4322
- [ ] #4324
- [ ] #4332
- [ ] #4337
- [ ] #4340
- [ ] #4341
- [ ] #4343
- [ ] #4355
- [ ] #4357

**Comment by liuzhe-lz** [comment]
## Bug Bash ##

### NAS

- [ ] @ultmaster SPOS
- [ ] @ultmaster ProxylessNAS
- [ ] @QuanluZhang Serialization
- [ ] @liuzhe-lz NAS visualization

### Model Compression

- [ ] @JiahangXu Movement pruning
- [ ] @JiahangXu AutoCompress Pruner
- [ ] @ultmaster AMC Pruner
- [ ] @QuanluZhang ptimizer & lr scheduler construct helper
- [ ] @J-shang v2 pruning example

### HPO

- [ ] @J-shang TPE
- [ ] @J-shang Grid Search

### nnictl

- Create experiment
  - [ ] @QuanluZhang v2 correct config
  - [ ] @QuanluZhang v2 wrong config
  - [ ] @QuanluZhang v1 correct config
  - [ ] @QuanluZhang v1 wrong config (will suggest user to manually convert)
  - [ ] @QuanluZhang annotation correct config
  - [ ] @QuanluZhang annotation wrong config
  - [ ] @J-shang From Python correct
  - [ ] @J-shang From Python wrong
  - [ ] @J-shang remote
  - [ ] @J-shang hybrid
  - [ ] @SparkSnail AML
- [ ] @QuanluZhang Using invalid field
- [ ] @QuanluZhang Report error when set trialGpuNumber to zero/undefined and maxTrialNumberPerGpu > 1

### Training Service

- @liuzhe-lz Hybrid for all reuse training services

### Web UI

- @liuzhe-lz Style update

**Comment by liuzhe-lz** [comment]
- [ ] To be fixed: search space not support 1e-5
- [ ] Torch lightning

**Comment by liuzhe-lz** [comment]
- [ ] #4206
- [ ] #4243
- [ ] #4280
- [ ] #4304
- [ ] #4305
- [ ] #4307
- [ ] #4308
- [ ] #4320
- [ ] #4321
- [ ] #4322
- [ ] #4328
- [ ] #4332
- [ ] #4337
- [ ] #4338
- [ ] #4341

- [ ] #4385
- [ ] #4390
- [ ] #4400
- [ ] #4412
- [ ] #4418
- [ ] #4420
- [ ] #4423
- [ ] #4428
- [ ] #4437
- [ ] #4438
- [ ] #4440
- [ ] #4445
- [ ] #4451
- [ ] #4457

**Comment by J-shang** [comment]
## Model Compression ##

+ New Pruner Supported in Pruning V2
  - Auto-Compress Pruner
  - AMC Pruner
  - Movement Pruning Pruner
+ Support `nni.trace` wrapped `Optimizer` in Pruning V2. In the case of not affecting the user experience as much as possible, trace the input parameters of the optimizer.
+ Optimize Taylor Pruner, APoZ Activation Pruner, Mean Activation Pruner in V2 memory usage.
+ Add more examples for Pruning V2
+ Add document for pruning config list
+ Bug Fix
  - Fix Slim Pruner in V2 not sparsify the BN weight.
  - Fix Simulator Annealing Task Generator generates config ignoring 0 sparsity.

**Comment by ultmaster** [comment]
### NAS

* Enhancement to serialization utilities (doc TBD) and changes to recommended practice of customizing evaluators (doc TBD)

### Bug fixes

* Fix a bug that causes DARTS 2nd order to crash.
* Fix a bug that causes deep copy of mutation primitives (e.g., LayerChoice) to crash.

**Comment by JiahangXu** [comment]
### NAS
- Refactor NAS examples of ProxylessNAS and SPOS.
- Support latency constraint on edge device for ProxylessNAS based on nn-Meter.
